src/util/LazyJsonParser.cpp

```diff
@@ -0,0 +1,258 @@
+      case '"':
+        parseLiteral(idx);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// ____________________________________________________________________________
+void LazyJsonParser::parseInArrayPath(size_t& idx, size_t& materializeEnd) {
+  AD_CORRECTNESS_CHECK(std::holds_alternative<InArrayPath>(state_));
+  auto& state = std::get<InArrayPath>(state_);
+
+  for (; idx < input_.size(); ++idx) {
+    if (idx - std::max(state.arrayStartIdx_, materializeEnd) >= 1'000'000) {
+      throw std::runtime_error(
+          "Ill-formed JSON: Element size exceeds 1 Megabyte.");
+    }
+    switch (input_[idx]) {
+      case '{':
+      case '[':
+        ++state.openBracketsAndBraces_;
+        break;
+      case '}':
+        --state.openBracketsAndBraces_;
+        break;
+      case ']':
+        if (state.openBracketsAndBraces_ == 0) {
+          // End of ArrayPath reached.
+          state_ = AfterArrayPath{.remainingBraces_ = arrayPath_.size(),
+                                  .arrayEndIdx_ = idx};
+          ++idx;
+          return;
+        }
+        --state.openBracketsAndBraces_;
+        break;
+      case ',':
+        if (state.openBracketsAndBraces_ == 0) {
+          materializeEnd = idx;
+          // reset arrayStartIdx_ to its sentinel 0, as it's only needed/valid
+          // for the first element
+          state.arrayStartIdx_ = 0;
+        }
+        break;
+      case '"':
+        parseLiteral(idx);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// ____________________________________________________________________________
+void LazyJsonParser::parseAfterArrayPath(size_t& idx, size_t& materializeEnd) {
+  AD_CORRECTNESS_CHECK(std::holds_alternative<AfterArrayPath>(state_));
+  auto& state = std::get<AfterArrayPath>(state_);
+
+  for (; idx < input_.size(); ++idx) {
+    if (idx - state.arrayEndIdx_ > 1'000'000) {
+      throw std::runtime_error(
+          "Ill-formed JSON: Element size exceeds 1 Megabyte.");
+    }
+    switch (input_[idx]) {
+      case '{':
+        state.remainingBraces_ += 1;
+        break;
+      case '}':
+        state.remainingBraces_ -= 1;
+        if (state.remainingBraces_ == 0) {
+          // End reached.
+          materializeEnd = idx + 1;
+        }
+        break;
+      case '"':
+        parseLiteral(idx);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// ____________________________________________________________________________
+std::string LazyJsonParser::constructResultFromParsedChunk(
+    size_t materializeEnd) {
+  std::string res;
+  if (materializeEnd == 0) {
+    return res;
+  }
+
+  // Don't yield before arrayPath.
+  AD_CORRECTNESS_CHECK(!std::holds_alternative<BeforeArrayPath>(state_));
+
+  if (yieldCount_ > 0) {
+    res = prefixInArray_;
+  }
+  ++yieldCount_;
+
+  absl::StrAppend(&res, input_.substr(0, materializeEnd));
+  input_ = input_.substr(std::min(materializeEnd + 1, input_.size()));
```

Comment:
This implicitly skips one character.
It is probably correct, but it looks really fishy.
This either needs documentation (why is this correct in all cases), or has to be fixed.
Is it correct , that this only happens either at a `,` inside the ArrayPath, or completely at the end after the trailer when the input is done.
So you can assert something like (inArrayPath && skippedCharIsComm) || ( afterArrayPath + materializeEnd is really at the end).